### PR TITLE
Feature/105360 cd remove hard coded user lists

### PR DESCRIPTION
**Pull request by chrisdexnimble**
**What is the change?**
Show list of available team members from database-data instead of hard-coded

**Why do we need the change?**
Needed to allow team members to be selected based on users known to the system

**What is the impact?**
Allows teams to be usable

**Azure DevOps Ticket**
105360

### CODE REVIEW

File: ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/HomePageModelTests.cs [hunk 1]
```diff
@@ -46,8 +46,8 @@ public async Task WhenInstanceOfIndexPageOnGetAsync_ReturnCases()
 			var defaultUsers = new[] { "user1", "user2" };
 			var roleClaimWrapper = new RoleClaimWrapper { Roles = roles, Users = defaultUsers };
 
-			mockRbacManager.Setup(r => r.GetUserRoleClaimWrapper(It.IsAny<string>()))
-				.ReturnsAsync(roleClaimWrapper);
+			//mockRbacManager.Setup(r => r.GetUserRoleClaimWrapper(It.IsAny<string>()))
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/HomePageModelTests.cs:49]: Would be good to know why this is commented out (normally I'd reject commented out code, but I guess this is because we're in a state of flux)

File: ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/HomePageModelTests.cs [hunk 2]
```diff
@@ -149,8 +149,8 @@ public async Task WhenInstanceOfIndexPageOnGetAsync_ReturnEmptyCases()
...
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/HomePageModelTests.cs:152]: Would be good to know why this is commented out (normally I'd reject commented out code, but I guess this is because we're in a state of flux)

File: ConcernsCaseWork/ConcernsCaseWork.Tests/Security/RbacManagerTests.cs [hunk 3]
```diff
@@ -21,8 +23,9 @@ public async Task WhenGetSystemUsers_Return_Users()
 			// arrange
 			var mockUserRoleCachedService = new Mock<IUserRoleCachedService>();
 			var mockLogger = new Mock<ILogger<RbacManager>>();
+			var mockTeamsService = new Mock<ITeamsCachedService>();
 			
-			var rbacManager = new RbacManager(BuildConfiguration(), mockUserRoleCachedService.Object, mockLogger.Object);
+			var rbacManager = new RbacManager(mockLogger.Object, mockTeamsService.Object);
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork.Tests/Security/RbacManagerTests.cs:28]: Not a blocker, but I'm not sure this test does what its title suggests - test that users are returned.

File: ConcernsCaseWork/ConcernsCaseWork/Security/RbacManager.cs [hunk 4]
```diff
@@ -13,44 +16,23 @@ namespace ConcernsCaseWork.Security
...
+		public async Task<IList<string>> GetSystemUsers(params string[] excludes)
 		{
-			_logger.LogInformation("RbacManager::GetUserRoleClaimWrapper {User}", user);
+			// TODO: Integrate the known users from the DB with Azure graph to build up a set of users where we can identify those who aren't in the graph and may have left.
 
-			var roleClaimWrapper = await _userRoleCachedService.GetRoleClaimWrapper(_defaultUsers, user);
-			
-			return roleClaimWrapper;
+			_logger.LogMethodEntered();
+			var users = await _teamsService.GetTeamOwners();
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork/Security/RbacManager.cs:34]: This is probably just me misunderstanding, but the method is called Get System Users (is that the users of the system, or system-level users? I guess the former). But it returns Team Owners. Is that all users? Or just a certain type of user? (I'm not sure if everyone is a team owner).  I'm also a bit unclear of the format of what goes into 'excludes' or the reason for it. Again, probably something that would be clearer when I've got more familiarity with the code, but thought I'd mention it.

File: ConcernsCaseWork/Service.Redis.Tests/Teams/TeamsCachedServiceTests.cs [hunk 5]
```diff
@@ -45,5 +42,92 @@ public void Constructors_GuardAgainstNullArgs()
 			var assertion = fixture.Create<GuardClauseAssertion>();
 			assertion.Verify(typeof(TeamsCachedService).GetConstructors());
 		}
+
+		[Test]
+		public async Task GetTeam_When_NotCached_Returns_Data_From_TeamsService()
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.Redis.Tests/Teams/TeamsCachedServiceTests.cs:47]: :heart:

File: ConcernsCaseWork/Service.Redis.Tests/Teams/TeamsCachedServiceTests.cs [hunk 6]
```diff
@@ -45,5 +42,92 @@ public void Constructors_GuardAgainstNullArgs()
...
+			var mockTeamsService = new Mock<ITeamsService>();
+			mockTeamsService.Setup(x => x.GetTeamOwners()).ReturnsAsync(new[] { expectedTeamOwner });
+
+			var sut = new TeamsCachedService(Mock.Of<ILogger<TeamsCachedService>>(), mockTeamsService.Object, mockCacheProvider.Object);
+
+			// act
+			var teamOwners = await sut.GetTeamOwners();
+
+			// assert
+			Assert.That(teamOwners.Length == 1);
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.Redis.Tests/Teams/TeamsCachedServiceTests.cs:102]: I'm not sure if Assert.Equal gives a better error message - I think it's recommended over using assert.that anyway. (or use It.Is format)

File: ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+			var fixture = new AutoFixture.Fixture();
+			fixture.Customize(new AutoMoqCustomization());
+			var assertion = fixture.Create<GuardClauseAssertion>();
+			assertion.Verify(typeof(TeamsService).GetConstructors());
+		}
+
+		[Test]
+		public async Task GetTeam_Returns_Dto()
+		{
+			var expectedDto = new ConcernsCaseworkTeamDto("user.one", new[] { "user.two" });
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs:44]: Again, a super small issue. Should we be using // arrange // act // assert? I don't mind either way, but nice to be consistent.

File: ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+			var responseWrapper = new ApiWrapper<ConcernsCaseworkTeamDto>(expectedDto);
+			var configuration = new ConfigurationBuilder().ConfigurationUserSecretsBuilder().Build();
+			var tramsApiEndpoint = configuration["trams:api_endpoint"];
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs:47]: I think this endpoint can be anything tbf - doesn't need to be pulled from config. I've noticed this elsewhere though. V. minor.

File: ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+				{
+					StatusCode = HttpStatusCode.NoContent
+				});
+
+			var httpClient = new HttpClient(mockMessageHandler.Object);
+			httpClient.BaseAddress = new Uri(tramsApiEndpoint);
+			httpClientFactory.Setup(_ => _.CreateClient(It.IsAny<string>())).Returns(httpClient);
+
+			var sut = new TeamsService(httpClientFactory.Object, Mock.Of<ILogger<TeamsService>>());
+			var result = await sut.GetTeam("user.one");
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs:88]: This is more of a comment on the implementation than the test, but would we consider returning an empty collection rather than a null? V. minor.

File: ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+			httpClientFactory.Setup(_ => _.CreateClient(It.IsAny<string>())).Returns(httpClient);
+
+			var sut = new TeamsService(httpClientFactory.Object, Mock.Of<ILogger<TeamsService>>());
+			var result = await sut.GetTeamOwners();
+
+			Assert.IsNotNull(result);
+			Assert.AreEqual(0, result.Length);
+		}
+
+		// Todo. Work out how to validate the Put method is working.
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs:148]: Has this been missed from this PR? (The Todo)

File: ConcernsCaseWork/Service.TRAMS/Base/AbstractService.cs [hunk 11]
```diff
@@ -1,17 +1,369 @@
...
+
+		public Task<T> Get<T>(string endpoint, bool treatNoContentAsError = false) where T : class 
+		{
+			Guard.Against.NullOrWhiteSpace(endpoint);
+
+			async Task<T> DoWork()
+			{
+				try
+				{
+					// Create a request
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.TRAMS/Base/AbstractService.cs:38]: I feel like we could strip some of these comments out of here :)

File: ConcernsCaseWork/Service.Redis/Teams/TeamsCachedService.cs [hunk 12]
```diff
@@ -24,7 +22,7 @@ public TeamsCachedService(ILogger<TeamsCachedService> logger, ITeamsService team
 		}
 
 		public Task<ConcernsCaseworkTeamDto> GetTeam(string ownerId)
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.Redis/Teams/TeamsCachedService.cs:24]: It's an interesting pattern you're using here - returning a task rather than executing it. Not one I've seen before. Would be good to know the background of it. I had to pull the branch to follow the code through to see that the code does actually get executed at some layer. Would be interesting to see if cancellation tokens are still used in this pattern too.

**Review by emma-w-nimble — APPROVED** [review]

File: ConcernsCaseWork/ConcernsCaseWork/Security/RbacManager.cs (hunk 4 above)
Comment (chrisdexnimble) [reply ConcernsCaseWork/ConcernsCaseWork/Security/RbacManager.cs:34]: That is all users. We are basically re-using the team owners as users who have been seen by the system.
(when they log in we are creating an empty team). 
We could do with refactoring it somehow but timescales are too tight.
So I hid this inside the RbacManager.

File: ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/HomePageModelTests.cs (hunk 1 above)
Comment (chrisdexnimble) [reply ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/HomePageModelTests.cs:49]: Not used anymore, roles are going to come from the bearer token, using the ClaimsPrincipalHelper. Will remove that comment.

File: ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/HomePageModelTests.cs (hunk 2 above)
Comment (chrisdexnimble) [reply ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/HomePageModelTests.cs:152]: Not used anymore, roles are going to come from the bearer token, using the ClaimsPrincipalHelper. Will remove that comment.

File: ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs (hunk 8 above)
Comment (chrisdexnimble) [reply ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs:47]: I did wonder that myself, will try it

File: ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs (hunk 9 above)
Comment (chrisdexnimble) [reply ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs:88]: I figured I'd let the service return a null, and then the caller use an empty collection. That way you'd know from the service whether you got nothing or really got an empty one.

File: ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs (hunk 10 above)
Comment (chrisdexnimble) [reply ConcernsCaseWork/Service.TRAMS.Tests/Teams/TeamsServiceTests.cs:148]: I couldn't work out how to tell if the put method endpoint was called. I tried to get it to return a task and assert the task was the same, but couldn't get it to work.

File: ConcernsCaseWork/Service.Redis/Teams/TeamsCachedService.cs (hunk 12 above)
Comment (chrisdexnimble) [reply ConcernsCaseWork/Service.Redis/Teams/TeamsCachedService.cs:24]: Cancellation tokens are still used. It allows fail-fast. More driven by how autofixture likes things in the idioms.
